src/main/java/net/minecraftforge/server/ForgeTimeTracker.java

```diff
@@ -22,108 +22,121 @@
 import java.lang.ref.WeakReference;
 import java.util.Arrays;
 import java.util.Map;
-import java.util.Map.Entry;
+import java.util.concurrent.TimeUnit;
 
-import com.google.common.collect.ImmutableMap;
-import com.google.common.collect.ImmutableMap.Builder;
+import com.google.common.collect.ImmutableList;
 import com.google.common.collect.MapMaker;
 
 import net.minecraft.entity.Entity;
 import net.minecraft.tileentity.TileEntity;
+import net.minecraftforge.server.timings.ForgeTimings;
 
-public class ForgeTimeTracker {
-    public static boolean tileEntityTracking;
-    public static int tileEntityTrackingDuration;
-    public static long tileEntityTrackingTime;
-    
-    private Map<TileEntity,int[]> tileEntityTimings;
-    private WeakReference<TileEntity> tile;
-    
-    private static final ForgeTimeTracker INSTANCE = new ForgeTimeTracker();
-
-    /* not implemented
-    private WeakReference<Entity> entity;
-    private Map<Entity,int[]> entityTimings;
-    */
-    
+/**
+ * A class to assist in the collection of data to measure the update times of ticking objects {currently Tile Entities and Entities}
+ *
+ * @param <T>
+ */
+public class ForgeTimeTracker<T>
+{
+
+    /**
+     * A tracker for timing tile entity update
+     */
+    public static ForgeTimeTracker<TileEntity> TILE_ENTITY_UPDATE = new ForgeTimeTracker<>();
+    /**
+     * A tracker for timing entity updates
+     */
+    public static ForgeTimeTracker<Entity> ENTITY_UPDATE = new ForgeTimeTracker<>();
+
+    public boolean enabled;
+    public int trackingDuration;
+    private Map<T, int[]> timings;
+    private WeakReference<T> currentlyTracking;
+    private long trackTime;
     private long timing;
-    
+
     private ForgeTimeTracker()
```

Comment:
Can this be made public so other mods may use the time tracker to track their own stuff?